### DISCUSSION TITLE
Undefined 'Model' on test with Karma Angular

### DISCUSSION THREAD

**Discussion by mpuerto-paymefy** [clientIO/joint#2615] [Q&A] [answered]
### Introduction

When I try to execute the tests with Karma in my Angular it says this:

![image](https://github.com/clientIO/joint/assets/163125788/305d1dc0-2dce-4fe6-920e-91fc1d44502d)

It seems the library JointJs it has a problem with the undefined "Model"

I only have two tests:

<img width="700" alt="image" src="https://github.com/clientIO/joint/assets/163125788/3f2fb369-b89c-4587-a634-4d1e630b2283">

I am working with:

![image](https://github.com/clientIO/joint/assets/163125788/f62b0b6d-9a9e-49c1-a816-83c0fe11af50)

### Steps to reproduce

`ng test`

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
Could you create a simple reproducible case (a repo)?

**Reply by mpuerto-paymefy** [reply]
Here the Repo: 

https://github.com/mpuerto-paymefy/model-error-lib-jointjs

When I `ng serve` and have my web in local you can also check the error in the console. This same error is the one from the Karma on `ng test`. The test are passing though.

**Reply by jamesgeorgewilliams** [reply]
Hi, is this the correct repository you meant to provide?

It's not possible to run `npm i` in this repository at the moment, because there are dependency conflicts.

It also seems like there are some tests/files missing. Currently, there are no `jointjs` related tests in the code provided.

In the picture above, the output mentions `AppComponent` and a test called `should create the app`, but none of this is present in the repository.

**Reply by mpuerto-paymefy** [reply]
I use `nom i -force`. 
Then the test is just the only test in the code.
In the package.json you can see the joint js v^3.7.2.
When you rung the test or the web it shows the same error

**Reply by jamesgeorgewilliams** [reply]
After running `npm i --force`, the following output is observed in your repository for `ng serve` and `ng test`.

`ng serve`:

```
./src/main.ts - Error: Module build failed (from ./node_modules/@ngtools/webpack/src/ivy/index.js):
Error: /Users/jameswilliams/Desktop/model-error-lib-jointjs/src/main.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.

./src/styles.scss - Error: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
Can't find stylesheet to import.
  ╷
8 │ @import '../../test-repo-fail-main/node_modules/primeflex/primeflex';
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
  src/styles.scss 8:9  root stylesheet

./src/styles.scss?ngGlobalStyle - Error: Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
HookWebpackError: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
Can't find stylesheet to import.
  ╷
8 │ @import '../../test-repo-fail-main/node_modules/primeflex/primeflex';
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
  src/styles.scss 8:9  root stylesheet

Error: error TS6053: File '/Users/jameswilliams/Desktop/test-repo-fail-main/src/main.ngtypecheck.ts' not found.
  The file is in the program because:
    Root file specified for compilation


Error: error TS6053: File '/Users/jameswilliams/Desktop/test-repo-fail-main/src/main.ts' not found.
  The file is in the program because:
    Root file specified for compilation
```

`ng test`:

```
./src/styles.scss - Error: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
Can't find stylesheet to import.
  ╷
8 │ @import '../../test-repo-fail-main/node_modules/primeflex/primeflex';
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
  src/styles.scss 8:9  root stylesheet

./src/styles.scss?ngGlobalStyle - Error: Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
HookWebpackError: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
Can't find stylesheet to import.
  ╷
8 │ @import '../../test-repo-fail-main/node_modules/primeflex/primeflex';
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
  src/styles.scss 8:9  root stylesheet

Error: error TS6053: File '/Users/jameswilliams/Desktop/test-repo-fail-main/src/test.ts' not found.
  The file is in the program because:
    Root file specified for compilation



22 04 2024 13:55:05.652:INFO [Chrome 124.0.0.0 (Mac OS 10.15.7)]: Connected on socket fx8pB8b1icEulGnAAAAB with id 64731660
22 04 2024 13:55:05.826:WARN [web-server]: 404: /_karma_webpack_/main.js
Chrome 124.0.0.0 (Mac OS 10.15.7): Executed 0 of 0 SUCCESS (0.027 secs / 0 secs)
TOTAL: 0 SUCCESS
```

Can you please check the repository you provided, and update it with functioning code?

**Reply by mpuerto-paymefy** [reply]
![image](https://github.com/clientIO/joint/assets/163125788/6c15aac9-5c32-468c-9036-e67c388838ac)

Here is the error in more detail
I am gonna update the repo.

**Reply by mpuerto-paymefy** [reply]
@jamesgeorgewilliams Updated the code, thanks for the patience.

**Reply by jamesgeorgewilliams** [reply, marked as answer]
Hi, thanks for updating the repository. 

If using a version of JointJS before version `4.0.0`, it is necessary to include the dependencies that JointJS had in the past.

Updating to the following in the `"test"` property in `angular.json` will fix the issue.

```js
"scripts": [
      "./node_modules/jquery/dist/jquery.min.js",
      "./node_modules/lodash/lodash.min.js",
      "./node_modules/backbone/backbone-min.js",
      "./node_modules/jointjs/dist/joint.min.js"
]
```

I strongly encourage you to update to the latest version of JointJS, `@joint/core` (`4.0.2`) . It is dependency free, and you won't run into this issue. 
https://www.npmjs.com/package/@joint/core
